File: src/lib/custom-deploy/custom-deploy.ts
import chalk from 'chalk';
import fs from 'fs';
import gql from 'graphql-tag';

import API from '../../lib/api';
import * as exit from '../../lib/cli/exit';
import { checkFileAccess, getFileSize, isFile, FileMeta } from '../../lib/client-file-uploader';
import { GB_IN_BYTES } from '../../lib/constants/file-size';
import { trackEventWithEnv } from '../../lib/tracker';
import {
	findLargeArchiveFilesInDeployArchive,
	LARGE_ARCHIVE_FILE_SIZE_LIMIT,
	validateDeployFileExt,
	validateFilename,
} from '../../lib/validations/custom-deploy';

import type { LargeArchiveFile } from '../../lib/validations/custom-deploy';

const DEPLOY_MAX_FILE_SIZE = 4 * GB_IN_BYTES;
const WPVIP_DEPLOY_TOKEN = process.env.WPVIP_DEPLOY_TOKEN;
const SKIP_LARGE_FILE_VALIDATION_FLAG = '--skip-large-file-validation';

type CustomDeployInfo = {
	success: boolean;
	appId: number;
	envId: number;
	envType: string;
	envUniqueLabel: string;
	primaryDomainName: string;
	launched: boolean;
};

type ValidateMutationPayload = {
	validateCustomDeployAccess: CustomDeployInfo;
};

function formatFileSize( bytes: number ): string {
	return `${ ( bytes / ( 1024 * 1024 ) ).toFixed( 1 ) } MB`;
}

function getLargeArchiveFilesMessage( largeArchiveFiles: LargeArchiveFile[] ): string {
	const fileList = largeArchiveFiles
		.map( file => `- ${ file.path } (${ formatFileSize( file.size ) })` )
		.join( '\n' );

	return `Deploy archive contains archive file(s) larger than ${ formatFileSize(
		LARGE_ARCHIVE_FILE_SIZE_LIMIT
	) } in the deploy archive:\n${ fileList }\nRemove these files from the repository, or rerun with ${ chalk.bold(
		SKIP_LARGE_FILE_VALIDATION_FLAG
	) } to skip this check.`;
}

export async function validateCustomDeployKey(
	app: string | number,
	env: string | number
): Promise< CustomDeployInfo > {
	if ( ! WPVIP_DEPLOY_TOKEN ) {
		exit.withError( 'Valid custom deploy key is required.' );
	}

	const VALIDATE_CUSTOM_DEPLOY_ACCESS_MUTATION = gql`
	mutation ValidateCustomDeployAccess {
		validateCustomDeployAccess( input: { app: "${ String( app ) }", env: "${ String( env ) }" } ) {
			success,
			appId,
			envId,
			envType,
			envUniqueLabel,
			primaryDomainName,
			launched
		}
	}
`;

	const api = API( { exitOnError: true } );
	try {
		const result = await api.mutate< ValidateMutationPayload >( {
			mutation: VALIDATE_CUSTOM_DEPLOY_ACCESS_MUTATION,
			context: {
				headers: {
					Authorization: `Bearer ${ WPVIP_DEPLOY_TOKEN }`,
				},
			},
		} );

		if ( ! result.data?.validateCustomDeployAccess ) {
			throw new Error( 'Not found' );
		}

		return result.data?.validateCustomDeployAccess;
	} catch {
		exit.withError( `Unauthorized: Invalid or non-existent custom deploy key for environment.` );
	}
}

/**
 * @param {FileMeta} fileMeta
 */
export async function validateFile( appId: number, envId: number, fileMeta: FileMeta ) {
	const { fileName, basename, isCompressed } = fileMeta;
	const track = trackEventWithEnv.bind( null, appId, envId );

	if ( ! fs.existsSync( fileName ) ) {
		await track( 'deploy_app_command_error', { error_type: 'invalid-file' } );
		exit.withError( `Unable to access file ${ fileMeta.fileName }` );
	}

	if ( ! isCompressed ) {
		await track( 'deploy_app_command_error', { error_type: 'uncompressed-file' } );
		exit.withError( `Please compress file ${ fileMeta.fileName } before uploading.` );
	}

	try {
		validateFilename( basename );
	} catch ( error ) {
		await track( 'deploy_app_command_error', { error_type: 'invalid-filename' } );
		exit.withError( error as Error );
	}

	try {
		validateDeployFileExt( fileName );
	} catch ( error ) {
		await track( 'deploy_app_command_error', { error_type: 'invalid-extension' } );
		exit.withError( error as Error );
	}

	try {
		await checkFileAccess( fileName );
	} catch {
		await track( 'deploy_app_command_error', { error_type: 'appfile-unreadable' } );
		exit.withError( `File '${ fileName }' does not exist or is not readable.` );
	}

	if ( ! ( await isFile( fileName ) ) ) {
		await track( 'deploy_app_command_error', { error_type: 'appfile-notfile' } );
		exit.withError( `Path '${ fileName }' is not a file.` );
	}

	const fileSize = await getFileSize( fileName );
	if ( ! fileSize ) {
		await track( 'deploy_app_command_error', { error_type: 'appfile-empty' } );
		exit.withError( `File '${ fileName }' is empty.` );
	}

	if ( fileSize > DEPLOY_MAX_FILE_SIZE ) {
		await track( 'deploy_app_command_error', {
			error_type: 'appfile-toobig',
			file_size: fileSize,
		} );
		exit.withError(
			`The deploy file size (${ fileSize } bytes) exceeds the limit (${ DEPLOY_MAX_FILE_SIZE } bytes).`
		);
	}
}

/**
 * @param {number} appId
 * @param {number} envId
 * @param {FileMeta} fileMeta
 */
export async function validateLargeArchiveFiles(
	appId: number,
	envId: number,
	fileMeta: FileMeta
) {
	const track = trackEventWithEnv.bind( null, appId, envId );
	let largeArchiveFiles: LargeArchiveFile[];

	try {
		largeArchiveFiles = await findLargeArchiveFilesInDeployArchive( fileMeta.fileName );
	} catch ( error ) {
		await track( 'deploy_app_command_error', {
			error_type: 'large-archive-file-verify-failed',
			verify_error: ( error as Error ).message,
		} );

		return exit.withError(
			`Unable to verify large archive files in the deploy archive: ${
				( error as Error ).message
			}. Rerun with ${ SKIP_LARGE_FILE_VALIDATION_FLAG } to skip this check.`
		);
	}

	if ( ! largeArchiveFiles.length ) {
		return;
	}

	await track( 'deploy_app_command_error', {
		error_type: 'large-archive-files',
		large_archive_files: largeArchiveFiles.map( file => file.path ),
	} );

	return exit.withError( getLargeArchiveFilesMessage( largeArchiveFiles ) );
}

File: internal/customdeploy/customdeploy.go
// Package customdeploy ports src/lib/custom-deploy/custom-deploy.ts and
// src/lib/validations/custom-deploy.ts — the gates and archive checks
// behind `vip app deploy` (+ `validate`).
package customdeploy

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/Automattic/vip/internal/upload"
)

// DeployMaxFileSize — DEPLOY_MAX_FILE_SIZE = 4 GiB (custom-deploy.ts:11).
const DeployMaxFileSize = int64(4) * 1024 * 1024 * 1024

// DeployInfo mirrors CustomDeployInfo (custom-deploy.ts:14).
type DeployInfo struct {
	AppID             int64
	EnvID             int64
	EnvType           string
	EnvUniqueLabel    string
	PrimaryDomainName string
	Launched          bool
}

// validFilenameRE — validations/custom-deploy.ts:49 (same charset as
// import sql, different message).
var validFilenameRE = regexp.MustCompile(`(?i)^[a-z0-9\-_.]+$`)

// ValidateDeployFilename ports validateFilename (validations/custom-deploy.ts:48).
func ValidateDeployFilename(filename string) error {
	if !validFilenameRE.MatchString(filename) {
		return fmt.Errorf("Filename %s contains disallowed characters: [0-9,a-z,A-Z,-,_,.]", filename)
	}
	return nil
}

// ValidateDeployFileExt ports validateDeployFileExt
// (validations/custom-deploy.ts:31): .zip, .tar.gz, or .tgz.
func ValidateDeployFileExt(filename string) error {
	ext := strings.ToLower(filepath.Ext(filename))
	if ext == ".gz" && strings.ToLower(filepath.Ext(strings.TrimSuffix(filename, filepath.Ext(filename)))) == ".tar" {
		ext = ".tar.gz"
	}
	if ext != ".zip" && ext != ".tar.gz" && ext != ".tgz" {
		return errors.New("Invalid file extension. Please provide a .zip, .tar.gz, or a .tgz file.")
	}
	return nil
}

// ValidateFile ports validateFile (custom-deploy.ts:74): the gate
// sequence ahead of upload. maxSize is injectable for tests; 0 uses the
// 4 GiB production limit.
func ValidateFile(meta upload.FileMeta, maxSize int64) error {
	if maxSize == 0 {
		maxSize = DeployMaxFileSize
	}

	fi, statErr := os.Stat(meta.FileName)
	if statErr != nil {
		return fmt.Errorf("Unable to access file %s", meta.FileName)
	}
	if !meta.IsCompressed {
		return fmt.Errorf("Please compress file %s before uploading.", meta.FileName)
	}
	if err := ValidateDeployFilename(meta.BaseName); err != nil {
		return err
	}
	if err := ValidateDeployFileExt(meta.FileName); err != nil {
		return err
	}
	if f, err := os.Open(meta.FileName); err != nil { // #nosec G304 -- checkFileAccess parity
		return fmt.Errorf("File '%s' does not exist or is not readable.", meta.FileName)
	} else {
		f.Close()
	}
	if fi.IsDir() {
		return fmt.Errorf("Path '%s' is not a file.", meta.FileName)
	}
	if fi.Size() == 0 {
		return fmt.Errorf("File '%s' is empty.", meta.FileName)
	}
	if fi.Size() > maxSize {
		return fmt.Errorf("The deploy file size (%d bytes) exceeds the limit (%d bytes).", fi.Size(), maxSize)
	}
	return nil
}
